File: src/main/java/com/ggardet/doctor/HospitalEventManager.java
package com.ggardet.doctor;

import com.ggardet.core.HospitalDeletedEvent;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.modulith.events.ApplicationModuleListener;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
@Slf4j
class HospitalEventManager {
    private final DoctorServiceInterface doctorServiceInterface;

    @ApplicationModuleListener
    void onRemovedHospitalEvent(final HospitalDeletedEvent event) {
        final var hospitalId = event.hospitalId();
        log.info("Received removal of hospital {}. Need to unlink doctors", hospitalId);
        final var doctors = doctorServiceInterface.findAllByHospital(hospitalId);
        doctors.forEach(doctorServiceInterface::unlinkDoctorFromHospital);
        log.info("Finished unlinking all doctors from hospital {}.", hospitalId);
    }
}

File: src/main/kotlin/com/ggardet/doctor/HospitalEventManager.kt
package com.ggardet.doctor

import com.ggardet.core.HospitalDeletedEvent
import org.slf4j.LoggerFactory
import org.springframework.modulith.events.ApplicationModuleListener
import org.springframework.stereotype.Service

@Service
internal class HospitalEventManager(
    private val doctorServiceInterface: DoctorServiceInterface
) {
    private val log = LoggerFactory.getLogger(HospitalEventManager::class.java)

    @ApplicationModuleListener
    fun onRemovedHospitalEvent(event: HospitalDeletedEvent) {
        val hospitalId = event.hospitalId
        log.info("Received removal of hospital {}. Need to unlink doctors", hospitalId)
        val doctors = doctorServiceInterface.findAllByHospital(hospitalId)
        doctors.forEach { doctor -> doctorServiceInterface.unlinkDoctorFromHospital(doctor) }
        log.info("Finished unlinking all doctors from hospital {}.", hospitalId)
    }
}